### ISSUE TITLE
FunctionType should override hashCode/equals?

### ISSUE THREAD

**Issue by fengguangyuan** [trinodb/trino#15175]
Hey, guys. I want to report an issue of `Metaspace OOM` running this SQL:
```sql
SELECT *
FROM (
  SELECT ARRAY['content'] as keys, MAP(ARRAY['A`contented'], ARRAY[ARRAY[1, 2]]) as mvalues
) t1
WHERE cardinality(map_filter(mvalues, (key, value) ->
      CASE WHEN split(key, '`') [array_position(keys,'consented')] = 'consented'
                AND cardinality(value) > 0 THEN TRUE
      ELSE FALSE
      END)) > 0);
```
This sql will cause the critical pressure on 1.4GB Metaspace with JDK11 & G1GC, after dug it, I noticed that `PageFunctionCompiler` will try to load cache entry using the `RowExpression` as the key translated from the filter predicate, but it always fails on matching the same row expression, because of containing a `BindExpression`. 
The `BindExpression` represents the lambda expression in the predicate, of which first argument is a `captured variable` and returned type is a `FunctionType`.
However `FunctionType` does not override hashCode/equals, it's impossible to generate the same hash code even for the same type signature.

In short,  if a row expression contains a function type, the cache is invalid, leading to the Metaspace under pressure, so should `FunctionType` override hashCode/equals?

**Comment by findepi** [comment]
cc @dain @martint

**Comment by dain** [comment]
Current JDK for Trino is 26.  JDK 11 has issues with metaspace, and I believe it has been completely removed from the JVMs no.

**Comment by findepi** [comment]
> However `FunctionType` does not override hashCode/equals

it now does -- https://github.com/trinodb/trino/blob/75ef0b43712a9de788ac1da944c6bbf91d65f8eb/core/trino-main/src/main/java/io/trino/type/FunctionType.java#L231-L248
